sqlite/sqlite.go

```diff
@@ -81,23 +82,33 @@ func openDB(path string, key string, kdfIterationsNumber int) (*sql.DB, error) {
 		return nil, err
 	}
 
-	// Disable concurrent access as not supported by the driver
-	db.SetMaxOpenConns(1)
-
-	if _, err = db.Exec("PRAGMA foreign_keys=ON"); err != nil {
-		return nil, err
-	}
-	keyString := fmt.Sprintf("PRAGMA key = '%s'", key)
-	if _, err = db.Exec(keyString); err != nil {
-		return nil, errors.New("failed to set key pragma")
-	}
-
-	if kdfIterationsNumber <= 0 {
-		kdfIterationsNumber = sqlite.ReducedKDFIterationsNumber
-	}
-
-	if _, err = db.Exec(fmt.Sprintf("PRAGMA kdf_iter = '%d'", kdfIterationsNumber)); err != nil {
-		return nil, err
+	// Workaround: casting the driver type breaks the database/sql abstraction
+	// and may lead to compatibility issues in the future.
+	// This method is used because the 'database/sql' package does not expose 'ConnectHook',
+	// thereby making it impossible to individually configure each connection.
+	// Consequently, the connections from the pool can't be properly decrypted, making them unusable.
+	sqlcipherDriver, ok := db.Driver().(*sqlcipher.SQLiteDriver)
```

Comment:
In my tests I found the best performance when using a limited amount of connections and the same number of max idle connections. The peak performance I see on M1 Macs is when using:
```
	db.SetMaxOpenConns(10)
	db.SetMaxIdleConns(10)
```
Not sure it's a coincidence to be the same number as CPU count.

Leaving the max open connection to unlimited will open a new connection when it's needed, and close the idle connections. This has the potential of paying the cost of opening the connection on lots of queries.

The difference I have is (In the app the difference might be smaller because we're probably not throttling the DB at this magnitude, but it gives an idea on how the connection pool works):

32 queries/second with unlimited connections
60662 queries/second with 10 max connections, 10 max idle connections

Reply:
Something like this should be used: https://go.dev/play/p/IFnlTmTD6Q

Reply:
I've done some benchmarking and it seems for my machine the number of `maxIdleConns` doesn't really matter:

![image](https://github.com/status-im/status-go/assets/33099791/d341e0f5-c414-4aa7-9f55-3bc30e843a5f)

@alexjba could you please give it a try on M1? Benchmarking code: https://github.com/osmaczko/status-go-db-perf/

Reply:
I set `maxOpenConns` to `nproc` as suggested and left `maxIdleConns` as default = 2 (see comment above).